### PR TITLE
Added percentageVisible{Max,Min} conditions for visibilitySpec.

### PR DESCRIPTION
**Pull request by avimehta**
Part of #1297

Tests to follow.

### CODE REVIEW

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 1]
```diff
@@ -217,15 +218,28 @@ export class InstrumentationService {
       return;
     }
 
-    // TODO(avimehta, #1297): Add all the listeners so that visibility
-    // conditions are monitored and callback is called when the conditions
-    // are met.
+    if (!config['visibilitySpec']) {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/instrumentation.js]: Mind flipping the conditionals? `!else` double-negatives make everything confusing.

File: src/service/visibility-impl.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    this.boundScrollListener_ = null;
+  }
+
+  /**
+   * @param {!JSONObject} config
+   * @param {!VisibilityListenerDef} callback
+   */
+  listenOnce(config, callback) {
+    const elId = config['selector'];
+    const res = resourcesFor(this.win_).get().find(el => {
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: A `document.getElementById` might be faster here. Also, we can attach `resources`:

```js
const element = this.doc_.getElementById(elId);
const resource = this.resources_.getResourceForElement(element);
//...
```

File: src/service/visibility-impl.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      return el.element.id == elId.slice(1);
+    });
+
+    if (!this.boundScrollListener_) {
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: Why not track with a boolean and bind `boundScrollListener_` in the constructor?

File: src/service/visibility-impl.js (hunk 2 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: Still looking through feedback and have not looked into this in depth but doesn't [getResourceForElement](https://github.com/ampproject/amphtml/blob/master/src/service/resources-impl.js#L316) expect an `AmpElement`?

File: src/service/visibility-impl.js (hunk 3 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: will do. Any reason it is preferable to what I am doing here?

Comment (jridgewell) [inline src/service/visibility-impl.js]: Just personal preference. :grin:

File: src/service/visibility-impl.js (hunk 2 above)
Comment (jridgewell) [inline src/service/visibility-impl.js]: I think selector is on an `AmpElement`?

**Comment by avimehta** [comment]
ptal. I am planning to add a few more tests but other stuff is ready for review.

File: src/service/visibility-impl.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * This type signifies a callback that gets called when visibility conditions
+ * are met.
+ * @typedef {function()}
+ */
+let VisibilityListenerCallbackDef;
+
+/**
+ * @typedef {Object.<string, JSONObject|VisibilityListenerCallbackDef>}
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: This always seems to be a `VisibilityListenerDef`.

File: src/service/visibility-impl.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  constructor(win) {
+    this.win_ = win;
+
+    /**
+     * key: resource id.
+     * value: [{ config: <config>, callback: <callback>}]
+     * @type {Object.<string, Array.<VisibilityListenerDef>>}
+     * @private
+     */
+    this.listeners_ = {};
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: `Object.create(null)`, please.

File: src/service/visibility-impl.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+export class Visibility {
+
+  /** @param {!Window} */
+  constructor(win) {
+    this.win_ = win;
+
+    /**
+     * key: resource id.
+     * value: [{ config: <config>, callback: <callback>}]
+     * @type {Object.<string, Array.<VisibilityListenerDef>>}
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: We don't use `.` inside type definitions.

File: src/service/visibility-impl.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  /**
+   * @param {!JSONObject} config
+   * @param {!VisibilityListenerDef} callback
+   */
+  listenOnce(config, callback) {
+    const element = this.win_.document.getElementById(config['selector']
+        .slice(1));
+    const res = this.resourcesService_.getResourceForElement(element);
+
+    if (!this.scrollListenerRegistered__) {
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: This might be better in a separate method.

File: src/service/visibility-impl.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+      const viewport = viewportFor(this.win_);
+
+      // Currently unlistens are not being used. In the event that no resources
+      // are actively being monitored, the scrollListener should be very cheap.
+      viewport.onScroll(this.boundScrollListener_);
+      viewport.onChanged(this.boundScrollListener_);
+      this.scrollListenerRegistered_ = true;
+    }
+
+    this.listeners_[res.getId()] = (this.listeners_[res.getId()] || []);
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: Might be cleaner to store `const id = res.getId()`

File: src/service/visibility-impl.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  /** @private */
+  scrollListener_() {
+    for (let r = this.resources_.length - 1; r >= 0; r--) {
+      const res = this.resources_[r];
+      const change = res.element.getIntersectionChangeEntry();
+      const ir = change.intersectionRect;
+      const br = change.boundingClientRect;
+      const visible = ir.width * ir.height * 100 / (br.height * br.width);
+
+      for (let c = this.listeners_[res.getId()].length - 1; c >= 0; c--) {
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: Let's store `this.listeners_[res.getId()]` to a variable.

File: src/service/visibility-impl.js (hunk 4 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: Config is a json object. I add it to the object at https://github.com/ampproject/amphtml/pull/2881/files#diff-6936d113e81d9bf7788ab8e145841720R88

File: src/service/visibility-impl.js (hunk 6 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: ack.

File: src/service/visibility-impl.js (hunk 5 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: ack.

File: src/service/visibility-impl.js (hunk 7 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: done.

File: src/service/visibility-impl.js (hunk 8 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: ack.

File: src/service/visibility-impl.js (hunk 9 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: done.

File: src/service/visibility-impl.js (hunk 4 above)
Comment (jridgewell) [inline src/service/visibility-impl.js]: The `|` is the confusing part then. Can you create a `VisibilityListenerConfigDef`?

File: src/service/visibility-impl.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+   */
+  listenOnce(config, callback) {
+    const element = this.win_.document.getElementById(config['selector']
+        .slice(1));
+    const res = this.resourcesService_.getResourceForElement(element);
+    const resId = res.getId();
+
+    this.registerForViewportEvents_();
+
+    this.listeners_[resId()] = (this.listeners_[resId()] || []);
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: This'll error, `resId` isn't a function.

File: src/service/visibility-impl.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    this.listeners_[resId()].push({
+      config: config,
+      callback: callback,
+    });
+    this.resources_.push(res);
+
+    this.scrollListener_(); // Should this be async?
```
Comment (jridgewell) [inline src/service/visibility-impl.js]: > Should this be async?

Might be good. Registering a delay, then canceling if there is a registered delay in `#scrollListener_` should work well.

File: src/service/visibility-impl.js (hunk 10 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: Weird. I swear I ran the tests. will fix.

File: src/service/visibility-impl.js (hunk 11 above)
Comment (avimehta) [inline src/service/visibility-impl.js]: done. ptal.

**Comment by jridgewell** [comment]
LGTM.

**Pull request merged**